### DISCUSSION TITLE
Properties with generic return types and decorators

### DISCUSSION THREAD

**Discussion by neerajanil** [microsoft/pyright#7314] [Q&A] [answered]
I am seeing inconsistent behavior between `pyright` and `mypy` in how the types are being resolved for properties which have decorators and generic return types

Sample code :

```python
import functools
from abc import abstractmethod
from typing import Callable, Generic, Protocol, TypeVar


class Foo(Protocol):

    @abstractmethod
    def name(self) -> str: ...


class FooImpl(Foo):

    def name(self) -> str:
        return "FooImpl"

    def print_name(self) -> None:
        print(self.name)


_U = TypeVar("_U")


def simple_decorator(func: Callable[..., _U]) -> Callable[..., _U]:
    return func


_T = TypeVar("_T", bound=Foo)


class Bar(Generic[_T]):

    def __init__(self, foo: _T) -> None:
        self._foo = foo
        super().__init__()

    @property
    def inner_foo(self) -> _T:
        return self._foo

    @functools.cached_property
    def inner_foo1(self) -> _T:
        return self._foo

    @property
    @functools.lru_cache(maxsize=10)
    def inner_foo2(self) -> _T:
        return self._foo

    @property
    @simple_decorator
    def inner_foo3(self) -> _T:
        return self._foo


class BarImpl(Bar[FooImpl]):

    def print_inner_foo(self) -> None:
        self.inner_foo.print_name()  # This is working in pyright and mypy
        self.inner_foo1.print_name()  # This is working in pyright and mypy
        self.inner_foo2.print_name()  # This is showing an error in pyright but not in mypy
        self.inner_foo3.print_name()  # This is showing an error in pyright but not in mypy
```
In pyright the following errors are coming up:
```
61:25 - error: Cannot access member "print_name" for type "Foo*"
    Member "print_name" is unknown (reportGeneralTypeIssues)
62:25 - error: Cannot access member "print_name" for type "Foo*"
    Member "print_name" is unknown (reportGeneralTypeIssues)
```
Looks like in pyright the type of `inner_foo` and `inner_foo1` are resolved to `FooImpl` while the type of `inner_foo2` and `inner_foo3` are resolved to `_T@Bar`. Is this a bug in either `mypy` or `pyright` ? Or am I missing something that is obvious.

**Comment by erictraut** [comment, marked as answer]
I don't consider this a bug in pyright. The behavior you're seeing is intended. I haven't analyzed mypy's behavior as closely, but it's quite likely that it's not a bug in mypy either. Type checkers use a bunch of special-case logic to handle the `property` class. You're violating some of the assumptions in this logic with your code sample above.

Your `simple_decorator` is obscuring the signature of the method it is decorating. You should use a ParamSpec to preserve its signature. If you make this change, it works fine.

```python
_U = TypeVar("_U")
_P = ParamSpec("_P")

def simple_decorator(func: Callable[_P, _U]) -> Callable[_P, _U]:
    return func
```

The `functools.lru_cache` decorator likewise obscures the signature of the method it's decorating. This is an [intentional choice made by the maintainers of typeshed](https://github.com/python/typeshed/issues/11280). I understand their reasoning, but it seems to me that their approach causes more problems than it solves. If you have an opinion on that issue, feel free to comment in the linked thread.

**Reply by neerajanil** [reply]
FYI for anyone checking this thread in the future, after looking through the discussion linked above and other related bugs and discussions. The solution I ended up using can be found below (adapted from [here](https://github.com/python/mypy/issues/5107#issuecomment-529372406)). Thank you @erictraut for the in-depth answer.

```python
if TYPE_CHECKING:
    _T = TypeVar("_T")
    _P = ParamSpec("_P")

    def lru_cache(
        maxsize: Optional[int] = 128, typed: bool = False
    ) -> Callable[[Callable[_P, _T]], Callable[_P, _T]]: ...

else:
    from functools import lru_cache
```
I also found this [comment](https://github.com/python/mypy/issues/5107#issuecomment-1355954910) very informative as to why the issue exists and the tradeoffs of various solutions.
